xla/tests/onednn_softmax_test.cc

```diff
@@ -27,208 +32,177 @@ limitations under the License.
-            reshape.11 = f32[16,128]{1,0} reshape(broadcast.10)
-            broadcast.12 = f32[16,128,30522]{2,1,0} broadcast(reshape.11), dimensions={0,1}
-            subtract.13 = f32[16,128,30522]{2,1,0} subtract(Arg_0.1, broadcast.12)
-            exponential.14 = f32[16,128,30522]{2,1,0} exponential(subtract.13)
-            constant.2 = f32[] constant(0)
-            reduce.19 = f32[16,128]{1,0} reduce(exponential.14, constant.2), dimensions={2}, to_apply=region_1.15
-            reshape.20 = f32[16,128,1]{2,1,0} reshape(reduce.19)
-            broadcast.21 = f32[16,128,1]{2,1,0} broadcast(reshape.20), dimensions={0,1,2}
-            reshape.22 = f32[16,128]{1,0} reshape(broadcast.21)
-            broadcast.23 = f32[16,128,30522]{2,1,0} broadcast(reshape.22), dimensions={0,1}
-            ROOT divide.24 = f32[16,128,30522]{2,1,0} divide(exponential.14, broadcast.23)
-        }
-    )";
-
-  EXPECT_TRUE(RunAndCompare(hlo_string, ErrorSpec{1e-4, 1e-4}));
-  MatchOptimizedHlo(hlo_string, onednn_softmax_);
-}
+  // Test the numerical correctness as well as pattern matching witn
+  // OneDnnOpsRewriter pass
+  void TestSoftmax(std::string softmax_hlo_string, float atol, float rtol) {
+    // check for pattern match
+    TF_ASSERT_OK_AND_ASSIGN(std::unique_ptr<HloModule> module,
+                            ParseAndReturnVerifiedModule(softmax_hlo_string));
+    OneDnnOpsRewriter softmax_rewrite_pass;
+    TF_ASSERT_OK_AND_ASSIGN(
+        bool changed, this->RunHloPass(&softmax_rewrite_pass, module.get()));
+    EXPECT_TRUE(changed);
+    EXPECT_THAT(module->entry_computation()->root_instruction(),
+                GmockMatch(::xla::match::CustomCall({"__onednn$softmax"})));
+
+    // check numerical correctness
+    EXPECT_TRUE(RunAndCompare(softmax_hlo_string, ErrorSpec{atol, rtol}));
+    MatchOptimizedHlo(softmax_hlo_string, onednn_softmax_);
+  }
+};
 
-TEST_F(OneDnnSoftmaxTest, SoftmaxFP32) {
-  const std::string hlo_string = R"(
-        HloModule jit_softmax, entry_computation_layout={(f32[1,128,30522]{2,1,0})->f32[1,128,30522]{2,1,0}}, allow_spmd_sharding_propagation_to_output={true}
-        region_0.4 {
-            Arg_0.5 = f32[] parameter(0)
-            Arg_1.6 = f32[] parameter(1)
-            ROOT maximum.7 = f32[] maximum(Arg_0.5, Arg_1.6)
+// Softmax test with last dimension as axis. In this case, axis = 2
+TEST_P(OneDnnSoftmaxTest, SoftmaxGenericTest) {
+  PrimitiveType data_type;
+  int batch_size;
+  std::tie(data_type, batch_size) = GetParam();
+  if (!IsSupportedType(data_type)) {
+    GTEST_SKIP() << "CPU does not support "
+                 << primitive_util::LowercasePrimitiveTypeName(data_type);
+  }
+  float atol = (data_type == F32) ? 1e-4 : 1e-2;
+  float rtol = (data_type == F32) ? 1e-4 : 1e-2;
+
+  const std::string softmax_hlo_template_string = R"(
+        HloModule softmax_module, entry_computation_layout={($0[$1,128,30522]{2,1,0})->$0[$1,128,30522]{2,1,0}}
```

Comment:
nit: here and in other tests, I don't think the `entry_computation_layout` needs to be specified. Please drop it if unnecessary for the sake of brevity :)

Reply:
Done